test/common/testlib.py

```diff
@@ -1439,19 +1507,55 @@ def test_main(options=None, suite=None, attachments=None, **kwargs):
         os.makedirs(opts.attachments, exist_ok=True)
 
     import __main__
+
+    def leaf_classes(base_class, exclude=None):
+        subclasses = base_class.__subclasses__()
+        if base_class == exclude:
+            return []
+        if not subclasses:
+            return [base_class]
+        return filter(lambda x: x.__module__ == "__main__", list(itertools.chain(*[leaf_classes(c, exclude) for c in subclasses])))
```

Comment:
this could probably be simplified with a list comprehension using an `if`? At the very least, why does it need the explicit `list()`? `filter()` works on iterables anyway.

Reply:
ah I think the inner list might have been a leftover from before, I added the `filter()` afterwards. I'll change it.